### PR TITLE
fix(core): invalidate previous failed projects if they are excluded

### LINKED ISSUE #4002: Excluding project from affected:build makes the build fail for that project if cache exists for that project
## Current Behavior
With Nx 10.3.0, i run  `nx affected:build --all`, the command build every app and then exit with the following error about foo:
```
Running build for affected projects failed.
Failed projects: foo
```

It also creates `results.json` file into cache folder like this 
```
{
  "command": "build",
  "results": {
    ...
    "foo": false
  }
}
```
 Then i rerun `nx affected:build --all --exclude=foo` and it doesn't build. 
```
Running build for affected projects failed.
Failed projects: foo
```

## Expected Behavior
Command:
```
nx affected:build --all --exclude=foo 
```
Running this command builds all apps except for foo and exit with success.

### Environment
```
nx : Not Found
  @nrwl/angular : 10.3.0
  @nrwl/cli : 10.3.0
  @nrwl/cypress : 10.3.0
  @nrwl/eslint-plugin-nx : Not Found
  @nrwl/express : Not Found
  @nrwl/jest : 10.3.0
  @nrwl/linter : 10.3.0
  @nrwl/nest : 10.3.0
  @nrwl/next : Not Found
  @nrwl/node : 10.3.0
  @nrwl/react : Not Found
  @nrwl/schematics : Not Found
  @nrwl/tao : 10.3.0
  @nrwl/web : Not Found
  @nrwl/workspace : 10.3.0
  typescript : 4.0.3
```

### PR DESCRIPTION
**Pull request by choarau-norsys**
## Current Behavior
Project build that failed once also failed with `--exclude` option

## Expected Behavior
Project build didn't failed with `--exclude` option, they are not build at all

## Related Issue(s)
Fixes #4002

### CODE REVIEW

File: packages/workspace/src/command-line/affected.ts [hunk 1]
```diff
@@ -45,7 +45,12 @@ export async function affected(
...
   }
   const projects = parsedArgs.all ? projectGraph.nodes : affectedGraph.nodes;
-  const env = readEnvironment(nxArgs.target, projects);
+  const projectsNotExcluded = Object.keys(projects)
+    .filter((key) => !parsedArgs.exclude.includes(key))
+    .reduce((_p, key) => {
+      return projects[key];
+    }, {});
+  const env = readEnvironment(nxArgs.target, projectsNotExcluded);
   const affectedProjects = Object.values(projects)
```
Comment (vsavkin) [inline packages/workspace/src/command-line/affected.ts:54]: Could you pass projectsNotExcluded here?

**Comment by vsavkin** [comment]
Thank you for the PR!

The change is good. Left two small comments.

File: packages/workspace/src/command-line/affected.ts (hunk 1 above)
Comment (choarau-norsys) [reply packages/workspace/src/command-line/affected.ts]: i had a bad reduce, corrected. Sorry for that !

i also pass `projectsNotExcluded` in commit 5c80987027879222a0adc5fd6c41a86f2d91aa3d

File: packages/workspace/src/command-line/affected.ts [hunk 2]
```diff
@@ -45,8 +45,14 @@ export async function affected(
...
-  const affectedProjects = Object.values(projects)
+  const projectsNotExcluded = Object.keys(projects)
+    .filter((key) => !parsedArgs.exclude.includes(key))
+    .reduce((p, key) => {
+      p[key] = projects[key];
+      return p;
+    }, {} as Record<string, ProjectGraphNode>);
+  const env = readEnvironment(nxArgs.target, projectsNotExcluded);
+  const affectedProjects = Object.values(projectsNotExcluded)
     .filter((n) => !parsedArgs.exclude.includes(n.name))
```
Comment (vsavkin) [inline packages/workspace/src/command-line/affected.ts:56]: I think you can remove this line now cause the filtering is happening above

Comment (choarau-norsys) [reply packages/workspace/src/command-line/affected.ts]: done

**Comment by vsavkin** [comment]
@choarau-norsys sorry, the formatter failed.  Could you run `yarn format` and repush?

**Comment by choarau-norsys** [comment]
> @choarau-norsys sorry, the formatter failed. Could you run `yarn format` and repush?

done but CI failed again

**Comment by vsavkin** [comment]
@choarau-norsys rebasing should fix it, but I'll merge it. The failures are unrelated and are due to a changed version we use for e2e tests.

**Comment by vsavkin** [comment]
Thank you!

**Pull request merged**